core/src/test/java/org/elasticsearch/index/translog/TranslogTests.java

```diff
@@ -2178,4 +2182,137 @@ public void testRollGenerationBetweenPrepareCommitAndCommit() throws IOException
+                                e -> e.getValue().stream().max(Long::compareTo).orElse(Long.MAX_VALUE)));
+
+        for (long seqNo = 0; seqNo < operations; seqNo++) {
+            final long finalLongSeqNo = seqNo;
+            final Long operand =
+                    maxSeqNoByGeneration
+                            .entrySet()
+                            .stream()
+                            .filter(e -> finalLongSeqNo <= e.getValue())
+                            .map(Map.Entry::getKey).min(Long::compareTo)
+                            .orElse(Long.MIN_VALUE);
+            assertThat(translog.getMinGenerationForSeqNo(seqNo).translogFileGeneration, equalTo(operand));
+        }
+
+    }
+
+
+
+    public void testSimpleCommit() throws IOException {
+        final int operations = randomIntBetween(1, 4096);
+        long seqNo = 0;
+        for (int i = 0; i < operations; i++) {
+            translog.add(new Translog.NoOp(seqNo++, 0, "test'"));
+            if (rarely()) {
+                translog.rollGeneration();
+            }
+        }
+
+        final long generation =
+                randomIntBetween(1, Math.toIntExact(translog.currentFileGeneration()));
+        translog.commit(generation);
+        for (long i = 0; i < generation; i++) {
+            assertFileDeleted(translog, i);
+        }
+        for (long i = generation; i <= translog.currentFileGeneration(); i++) {
+            assertFileIsPresent(translog, i);
+        }
+    }
+
+    public void testPrepareCommitAndCommit() throws IOException {
+        final int operations = randomIntBetween(1, 4096);
+        long seqNo = 0;
+        for (int i = 0; i < operations; i++) {
+            translog.add(new Translog.NoOp(seqNo++, 0, "test"));
+            if (rarely()) {
+                final long generation = translog.currentFileGeneration();
+                translog.prepareCommit();
+                if (rarely()) {
+                    // simulate generation filling up and rolling between preparing the commit and committing
+                    translog.rollGeneration();
+                }
+                translog.commit(randomIntBetween(1, Math.toIntExact(generation)));
+                for (long g = 0; i < generation; g++) {
+                    assertFileDeleted(translog, g);
+                }
+                for (long g = generation; g < translog.currentFileGeneration(); g++) {
+                    assertFileIsPresent(translog, g);
+                }
+            }
+        }
+    }
+
+    public void testCommitWithOpenView() throws IOException {
+        final int operations = randomIntBetween(1, 4096);
+        long seqNo = 0;
+        long lastCommittedGeneration = -1;
+        for (int i = 0; i < operations; i++) {
+            translog.add(new Translog.NoOp(seqNo++, 0, "test"));
+            if (rarely()) {
+                if (randomBoolean()) {
+                    try (Translog.View ignored = translog.newView()) {
```

Comment:
With only the if branch, the view will always be taken from generation 1 to the last committed generation which means that these generations would never be trimmed. By having the else, sometimes earlier generations get trimmed (because there is no view to hold onto them) making the if branch a little more interesting.